### PR TITLE
feat: chunk handler

### PR DESCRIPTION
**Pull request by elijaharita**
Allows chunks to be created via clients.

Questions:
- Is `client.Chunk()` descriptive enough? Should I change it to something like `AddChunk()`?
- Is there anything wrong with the result in my test?
    `&model.Chunk{ID:0x1, CreatedAt:time.Date(2023, time.August, 3, 18, 12, 42, 764061047, time.Local), SourceID:0x1, Source:(*model.Source)(nil), PackingState:"ready", PackingWorkerID:(*string)(nil), PackingWorker:(*model.Worker)(nil), ErrorMessage:"", ItemParts:[]model.ItemPart(nil), Cars:[]model.Car(nil)}`

### CODE REVIEW

File: handler/datasource/pushitem.go [hunk 1]
```diff
@@ -84,3 +88,157 @@ func pushItemHandler(
...
+func MaxSizeToSplitSize(m int64) int64 {
+	r := util.NextPowerOfTwo(uint64(m)) / 4
+	if r > 1<<30 {
+		r = 1 << 30
+	}
+
+	return int64(r)
+}
+
+func ExtractFromFsObject(ctx context.Context, info fs.ObjectInfo) (size int64, hashValue string, lastModified time.Time, lastModifiedReliable bool) {
```
Comment (willscott) [inline handler/datasource/pushitem.go:224]: seems like this method should be able to return an error?

**Review by willscott — COMMENTED** [review]
looks generally good.

agree that the 'chunk' vs 'item' distinction is a bit hard to follow both in the code and in terms of feeling confident of which method i'd choose when interacting with the API

File: handler/datasource/pushitem.go (hunk 1 above)
Comment (elijaharita) [reply handler/datasource/pushitem.go:224]: i just moved it verbatim from another file, but i could add that functionality as a part of this pr if it would be useful

**Comment by elijaharita** [comment]
> looks generally good.
> 
> agree that the 'chunk' vs 'item' distinction is a bit hard to follow both in the code and in terms of feeling confident of which method i'd choose when interacting with the API

wait, what do u mean by this? i wasn't trying to make that point in this post, is that in response to a different discussion?

**Comment by willscott** [comment]
> > looks generally good.
> > agree that the 'chunk' vs 'item' distinction is a bit hard to follow both in the code and in terms of feeling confident of which method i'd choose when interacting with the API
> 
> wait, what do u mean by this? i wasn't trying to make that point in this post, is that in response to a different discussion?

In reading this PR, i found the plumbing through and exposure of both 'chunk' and 'item' terms at the API surface to be potentially confusing. Both are pretty generic terms that we are using in a specific way. If I was preparing a custom dataset against singularity, where would I learn about the semantic distinction and get a clear sense of which methods I should be calling? As we expose chunk handlers, it may be worth some extra comments in code and documentation to better explain how we're using these two terms.

**Comment by elijaharita** [comment]
> > > looks generally good.
> > > agree that the 'chunk' vs 'item' distinction is a bit hard to follow both in the code and in terms of feeling confident of which method i'd choose when interacting with the API
> > 
> > 
> > wait, what do u mean by this? i wasn't trying to make that point in this post, is that in response to a different discussion?
> 
> In reading this PR, i found the plumbing through and exposure of both 'chunk' and 'item' terms at the API surface to be potentially confusing. Both are pretty generic terms that we are using in a specific way. If I was preparing a custom dataset against singularity, where would I learn about the semantic distinction and get a clear sense of which methods I should be calling? As we expose chunk handlers, it may be worth some extra comments in code and documentation to better explain how we're using these two terms.

Ahh, I see what you mean. It does make the API less discoverable as we add more handlers that only need to be called in certain circumstances.

Maybe these "manual background" handlers I've been exposing should get their own directory? Something like /prep, and we can comment that they're usually handled by background workers?

On the other hand, we could leave them where they are but make it explicit that they don't usually need to be called by users?

**Comment by elijaharita** [comment]
this may be unnecessary. will update
